Add Language ID

J38 opened the pull request:
## Description
Adds language id functionality. Implements a multilingual pipeline that takes in text, identifies language, and routes text to appropriate language specific pipeline. Also adds lang field to Document.

## Unit test coverage
TO DO

## Known breaking changes/behaviors
This should just add additional functionality.

J38 commented:
The main issue to resolve at this point is incorporating the language id model into the resources framework. My current implementation has a "multilingual" language, and the only processor it has is the "langid" processor.

J38 commented:
I should note the code requires this in resources.json to work:

```
{'tokenize': {'ud': {'md5': '1234'}}, 'default_dependencies': {}, 'default_processors': {'langid': 'ud'}, 'langid': {'ud': {'md5': '1234'}}}
```

J38 commented:
That is under the key "multilingual"

AngledLuffa commented:
The first question that comes to mind about "multilingual" is what other
contexts would that apply to

For example, suppose we add a multilingual transformer layer.  Would it
make sense for that to be in the same category as the language id model?
Maybe it does, now that I think about it some.
What other possibilities are there?  Models which annotate a specific
subset of languages at the same time?  (Spanish-English, Chinese-English,
Hindi-English for example)  Would those also qualify as multilingual or
would we need some other way to include them?

At any rate, the resources.json file is produced
in stanza/resources/prepare_resources.py

AngledLuffa commented on stanza/models/lang_identifier.py:
```diff
@@ -0,0 +1,193 @@
+    parser.add_argument("--randomize-range", help="range of lengths to use when random sampling text", 
+                        type=randomize_range, default="5,20")
+    parser.add_argument("--save-name", help="where to save model", default=None)
+    parser.add_argument("--use-gpu", help="whether to use gpu", type=bool, default=True)
```
This does not lead to a corresponding `--use-gpu=False` option - perversely, `bool("False") == True`.  Instead, you want something like what the tokenizer model does, where it uses `torch.cuda.is_available()` as the default and has a `--cpu` argument

J38 replied:
What seems ideal to me is `--use-gpu=False` to denote not using the GPU, and as we always agree default should be "try to use the gpu". I can fix this with the wonderful distutils library:

```
>>> from distutils.util import strtobool
>>> strtobool("False")
0
```

J38 replied:
There should be one flag that handles this though, and I guess to me the one flag should be `--use-gpu`.

J38 replied:
And of course I need to add the `torch.cuda.is_available()` rather than default to True.

J38 replied:
Ok I just changed this to a `--use-cpu` flag and otherwise it will try to use gpu.

J38 replied:
If it can find it.

AngledLuffa commented on stanza/models/lang_identifier.py:
```diff
@@ -0,0 +1,193 @@
+    return args
+
+
+def randomize_range(range_list):
```
this name makes it sound like you're doing something with random numbers.  also, you probably want to check that it's exactly 2 integers with x < y

J38 replied:
Ok I'll rename and add check...

AngledLuffa commented on stanza/models/lang_identifier.py:
```diff
@@ -0,0 +1,193 @@
+
+
+def main(args=None):
+    args = parse_args(args=args)
```
maybe set the random seed here?  helps reproducibility a bit

J38 replied:
Ok sounds like a plan.

AngledLuffa commented on stanza/models/lang_identifier.py:
```diff
@@ -0,0 +1,193 @@
+    char_to_idx = {}
+    train_files = [f"{args.data_dir}/{x}" for x in os.listdir(args.data_dir) if "train" in x]
+    for train_file in train_files:
+        lines = open(train_file).read().split("\n")
```
with open(train_file) as fin:
will automatically close the file when the block exits and lets you iterate the lines

AngledLuffa commented on stanza/models/lang_identifier.py:
```diff
@@ -0,0 +1,193 @@
+        "lang_weights": train_data.lang_weights
+    }
+    load_model = bool(args.load_model)
+    if load_model:
```
`if args.load_model` works without the bool()

J38 replied:
Yeah there was no value to that, I don't know why I did that...

J38 replied:
I like to feel like the function is getting called with True or False, but probably less lines of code is a better principle to follow

AngledLuffa commented on stanza/models/lang_identifier.py:
```diff
@@ -0,0 +1,193 @@
+    load_model = bool(args.load_model)
+    if load_model:
+        trainer_config["load_model"] = args.load_model
+        print(f"{datetime.now()}\tLoading model from: {args.load_model}")
```
logging instead of print() would be more consistent with the rest of the models

J38 replied:
Great point, will switch to our logging

AngledLuffa commented on stanza/models/lang_identifier.py:
```diff
@@ -0,0 +1,193 @@
+    """
+    Helper that will determine corresponding log file (e.g. /path/to/demo.pt to /path/to/demo.json
+    """
+    model_suffix = file_path.split(".")[-1]
```
`os.path.splitext` accomplishes this.  `-1 * len()` should not be necessary - you can just do `-len()`, at least in situations where `os.path.splitext` doesn't exist

J38 replied:
Yeah I guess I should stop manipulating the string and get in the habit of using the proper file path manipulating libraries

AngledLuffa commented on stanza/models/langid/create_ud_data.py:
```diff
@@ -0,0 +1,186 @@
+"""
+Script for producing training/dev/test data from UD data
+"""
```
A brief explanation of what the data looks like would go a long way

J38 replied:
No prob, will add to comment, the format is `{"text": "Hello world.", "label": "en"}`

AngledLuffa commented on stanza/models/langid/create_ud_data.py:
```diff
@@ -0,0 +1,186 @@
+    args = parse_args(args=args)
+    if isinstance(args.languages, str):
+        args.languages = args.languages.split(",")
+    data_paths = [f"{args.save_path}/{data_split}.jsonl" for data_split in ["train", "dev", "test"]]
```
jsonl?

J38 replied:
That is format one json dict per line ...I've seen it (Jimmy, it's what the inspiration project did), so I thought that was a new standard

AngledLuffa commented on stanza/models/langid/create_ud_data.py:
```diff
@@ -0,0 +1,186 @@
+    """
+    all_sentences = []
+    with open(ud_file_path) as ud_file:
+        sentences = [x[9:] for x in ud_file.read().split("\n") if x.startswith("# text = ")]
```
might want to verify that `# text` is a thing that occurs in the file

J38 replied:
They used a formal library for processing UD files, but I didn't want to add a dependency. But might be good to make it a little more resilient than just assuming that works.

AngledLuffa commented on stanza/models/langid/create_ud_data.py:
```diff
@@ -0,0 +1,186 @@
+    """
+    all_sentences = []
+    with open(ud_file_path) as ud_file:
+        sentences = [x[9:] for x in ud_file.read().split("\n") if x.startswith("# text = ")]
```
also, i would be surprised if you couldn't just do `for x in ud_file`

and the all_sentences vs sentences seems like an unused distinction here

AngledLuffa commented on stanza/models/langid/create_ud_data.py:
```diff
@@ -0,0 +1,186 @@
+    Create window size chunks from a sentence, always starting with a word
+    """
+    windows = []
+    words = sentence.split(" ")
```
this is tricky on account of two random language issues i can think of

some languages (FR for example) specify numbers as

100 000 000

being 10^8 for example

and that is treated as one token

so this `sentence.split(" ")` will break up tokens like that (why not just split() btw?)

the other is of course VI, where tokens have spaces all the time

AngledLuffa commented on stanza/models/langid/create_ud_data.py:
```diff
@@ -0,0 +1,186 @@
+    """
+    if len(current_window) < min_window:
+        return False
+    if not re.search('[a-zA-Z]', current_window):
```
what's the benefit of doing this?  i suppose it eliminates a list of numbers which could have been from many different languages, such as if someone lists a hockey lineup by their jersey numbers: 14, 28, 93, 9, 53, ...  OTOH it seems like it would unnecessarily prevent this tool from being used on BG, ZH, RU, HI, ...

J38 replied:
This was from the original project which was only working on latin characters, but I removed this constraint for our general language identifier.

AngledLuffa commented on stanza/models/langid/create_ud_data.py:
```diff
@@ -0,0 +1,186 @@
+    """
+    if len(current_window) < min_window:
+        return False
+    if not re.search('[a-zA-Z]', current_window):
```
also granted this is not going to be run by users very often, but for a fixed `re` expression it's a good idea to compile it.  greatly improves the runtime

AngledLuffa commented on stanza/models/langid/create_ud_data.py:
```diff
@@ -0,0 +1,186 @@
+    line = line.translate(remove_digits)
+    words = line.split()
+    new_words = []
+    # Below fixes large I instead of l. Does not catch everything, but should also not really make any mistakes either
```
how often does this happen?  perhaps we could file a bug report on the underlying data

J38 replied:
Just a note a few of those functions are directly lifted (and credited) from https://github.com/AU-DIS/LSTM_langid/blob/main/src/dataset_creator.py

J38 replied:
For better or worse design choices so far have been about matching their paper.

J38 replied:
I'm trying to recreate their system's results, which depend a lot on very precise data handling.

AngledLuffa commented on stanza/models/langid/data.py:
```diff
@@ -0,0 +1,122 @@
+        """
+
+        # set up lang counts used for weights for cross entropy loss
+        lang_counts = [0 for _ in tag_index]
```
dunno if this is more convenient, but there is actually a `collections.Counter` class

AngledLuffa commented on stanza/models/langid/data.py:
```diff
@@ -0,0 +1,122 @@
+        examples = []
+        for data_file in data_files:
+            examples += [x for x in open(data_file).read().split("\n") if x.strip()]
+        random.shuffle(examples)
```
is this shuffle something we want if `randomize` is `False`?

AngledLuffa commented on stanza/models/langid/data.py:
```diff
@@ -0,0 +1,122 @@
+        # set up lang weights
+        most_frequent = max(lang_counts)
+        for idx in range(len(lang_counts)):
+            lang_counts[idx] = float(most_frequent) / float(lang_counts[idx])
```
float() should not be needed any more in newer python - 5 / 8 -> 0.625 as you would expect

J38 replied:
woah!

AngledLuffa commented on stanza/models/langid/data.py:
```diff
@@ -0,0 +1,122 @@
+        # set up lang weights
+        most_frequent = max(lang_counts)
+        for idx in range(len(lang_counts)):
+            lang_counts[idx] = float(most_frequent) / float(lang_counts[idx])
```
also, is this an opportunity for a list comprehension?

`lang_counts = [most_frequent / x for x in lang_counts]`

AngledLuffa commented on stanza/pipeline/langid_processor.py:
```diff
@@ -0,0 +1,113 @@
+"""
+Processor for determing language of text.
```
ning

AngledLuffa commented:
I see the benefit of that for the first iteration.  Not sure we want to
shackle ourselves to that in the long run.  Perhaps an option to limit
ourselves to languages a-zA-Z and to convert random I to l would be a good
addition

AngledLuffa commented on stanza/pipeline/multilingual.py:
```diff
@@ -0,0 +1,104 @@
+        lang_id_config: dict = None,
+        lang_configs: dict = None,
+        ld_batch_size: int = 64,
+        max_cache_size: int = 10,
```
default size of 10 is pretty ambitious in a world where cryptominers are buying every 3090 in sight :) 

:(

AngledLuffa commented on stanza/pipeline/multilingual.py:
```diff
@@ -0,0 +1,104 @@
+        # update pipeline cache
+        if lang not in self.pipeline_cache:
+            # clear least recently used lang from pipeline cache
+            if len(self.pipeline_cache) == self.max_cache_size:
```
fwiw this is making the actual max_cache_size one less than the given max, isn't it?

AngledLuffa commented on stanza/pipeline/multilingual.py:
```diff
@@ -0,0 +1,104 @@
+
+        # create language specific batches, store global idx with each doc
+        lang_batches = {}
+        for doc in docs_w_langid:
```
Nice!  I was going to suggest batching by language, but I see you've already done that

AngledLuffa commented on stanza/pipeline/multilingual.py:
```diff
@@ -0,0 +1,104 @@
+                self.lang_request_history.remove(lru_lang)
+            self.pipeline_cache[lang] = Pipeline(**self.lang_configs[lang])
+
+    def process(self, doc):
```
is there room for a lang= parameter (or different method) where the language is specified for a certain document?

J38 replied:
So we might have some documents are unknown and some aren't, and we should check if it's already been assigned a language?

J38 commented on stanza/models/langid/create_ud_data.py:
```diff
@@ -0,0 +1,186 @@
+
+def generate_examples(lang_id, list_of_files, splits=(0.8,0.1,0.1), min_window=10, max_window=50, 
+                      eval_length=10):
+    """
```
Yeah in the original project they did 5-fold analysis, and just dumped all of the train/dev/test from UD into one bucket and made 5 folds. So I've just been matching.

J38 commented:
I think the special cases should be handled properly now.  The two I see are the Nynorsk vs. Bokmaal and Simplified vs. Traditional Chinese. The current UD --> training data script should appropriately label those examples, and it can now differentiate between Traditional and Simplified Chinese.

J38 commented:
is_available() is the default ... it will try to use the gpu unless `--use-cpu` is specified.

J38 commented:
The model is built with fine-grained so we can route to the appropriate pipeline. Most language id systems do not differentiate "zh-hans" and "zh-hant", so for apples to apples comparison with other language id systems I added this option which will collapse "zh-hans" and "zh-hant" (or future situations) to "zh".

J38 commented:
I'll change the name to `merge-labels-for-eval`.

J38 commented:
Also just as a note, I need to use the `emoji` open source library to detect emojis. This can help with language id scores on tweets, so I added that to the setup.py

AngledLuffa commented:
> is_available() is the default ... it will try to use the gpu unless `--use-cpu` is specified.

I saw that... what I meant was the section could look something like this block, from `tagger.py`

```
    parser.add_argument('--cuda', type=bool, default=torch.cuda.is_available())
    parser.add_argument('--cpu', action='store_true', help='Ignore CUDA.')
```

AngledLuffa commented:
Good to know.  We can test that it is being caught correctly by the
installation - there's testpypi, for example.

J38 commented:
Ok I'll model it after tagger.py

J38 commented:
Isn't `--cuda` bad though?  For instance won't `bool("False")` yield True?

J38 commented:
I honestly think the current set up is the best. If you do nothing, it will try to use the GPU if possible. If you are running on a GPU machine and explicitly don't want it to use the GPU you can use the `--use-cpu` flag to route it towards using the CPU. I don't think it makes anything better to have 2 flags related to the same thing.

J38 commented:
Or is the point the user will never use the `--cuda` flag?

J38 commented:
I'll make the setting of `args.use_gpu` a one-liner, but I think having an option that is never supposed to be used is sub-optimal.
